scrypto/src/math/decimal.rs

```diff
@@ -170,6 +173,65 @@ impl Decimal {
                 }
             }
         }
+
+    }
+
+    pub fn powi(&self, exp: i32) -> Self {
+        let mut x: BigInt = self.0.clone().into();
+        if x.is_negative() {
+            panic!("powi is not supported for negative numbers");
+        }
+        let s = Self::SCALE;
+        let bytes2 = Self::BITS / 4;
+        let b: BigInt = 10i128.pow(s).into();
+        let mut n: BigInt = exp.abs().into();
+        let mut z: BigInt;
+        if x.clone() == BigInt::from(0i8) {
+            if n.clone() == Zero::zero() {
+                z = b.clone();
+            } else {
+                z = Zero::zero();
+            }
+        } else if x.clone() == b.clone() {
+            z = b.clone();
+        } else {
+            if n.clone() % 2i8 == Zero::zero() {
+                z = b.clone();
+            } else {
+                z = x.clone();
+            }
+            let half = b.clone() / 2i8;
+            n /= 2i8;
+            while n.clone() > Zero::zero() {
+                let xx = x.clone() * x.clone();
+                if xx.to_signed_bytes_le().len() > bytes2 {
+                    panic!("pow overflow: square of {} too large", x);
+                }
+                let xx_round = xx.clone() + half.clone();
+                if xx_round.to_signed_bytes_le().len() > bytes2 {
+                    panic!("pow overflow: sum of {} and {} too large", xx, half);
+                }
+                x = xx_round.clone() / b.clone();
+                if n.clone() % 2i8 != Zero::zero() {
+                    let zx = z.clone() * x.clone();
+                    if zx.to_signed_bytes_le().len() > bytes2 {
+                        panic!("pow overflow: product of {} and {} too large", z, x);
+                    }
+                    let zx_round = zx.clone() + half.clone();
+                    if zx_round < zx.clone() {
+                        panic!("pow overflow: sum of {} and {} too large", zx, half);
+                    }
+                    z = zx_round.clone() / b.clone();
+                }
+                n /= 2i8;
+            }
+        }   
+
+        if exp.is_negative() {
+            Self(big_int_to_i128(b.clone() * b / z))
+        } else {
+            Self(big_int_to_i128(z))
+        }
```

Comment:
Personally, I feel like there are too many intermediary steps and rounding, which makes it super hard to describe the behaviour of the `powi` function and hard for me to verify the correctness.

````
(a * 10 ^ -18) ^ exp = a ^ exp * 10 ^ (-18 * exp) 
                     = a ^ exp * 10 ^ (-18 * exp + 18) * 10 ^ -18
                     = a ^ exp / 10 ^ (18 * exp + 18) * 10 ^ -18
````

It seems to we just need to raise `a` to the power of `exp`, raise `10` to the power of `18 * exp + 18`, and do the division of the these two numbers, which requires one rounding only.